### ISSUE TITLE
Rule to prevent redundant `Promise.resolve` or `Promise.reject`

### ISSUE THREAD

**Issue by ForbesLindesay** [xjamundx/eslint-plugin-promise#11]
See eslint/eslint#5760

Rule to warn when wrapping a value returned from a Promise handlerFn in `Promise.resolve` or `Promise.reject`

```js
something.then(function () {
  // `Promise.resolve` is redundant, just use `return val;`
  return Promise.resolve(val);
});

// and

something.then(function () {
  // `Promise.reject` is redundant, just use `throw val;`
  return Promise.reject(val);
});
```

**Comment by xjamundx** [comment]
Sorry I missed this earlier. We can probably do something like this.

**Comment by xjamundx** [comment]
Tentative test suite
```
'use strict'

var rule = require('../rules/no-return-wrap')
var RuleTester = require('eslint').RuleTester
var rejectMessage = 'Expected throw instead of Promise.reject'
var resolveMessage = 'Avoid wrapping return values in Promise.resolve'
var ruleTester = new RuleTester()

ruleTester.run('no-return-wrap', rule, {
  valid: [

    // resolve and reject are sometimes okay
    'Promise.resolve(4).then(function(x) { return x })',
    'Promise.reject(4).then(function(x) { return x })',
    'Promise.resolve(4).then(function() {})',
    'Promise.reject(4).then(function() {})',

    // throw and return are fine
    'doThing().then(function() { return 4 })',
    'doThing().then(function() { throw 4 })',
    'doThing().then(null, function() { return 4 })',
    'doThing().then(null, function() { throw 4 })',
    'doThing().catch(null, function() { return 4 })',
    'doThing().catch(null, function() { throw 4 })',

    // other Promise.things are fine
    'doThing().then(function() { return Promise.all([a,b,c]) })',

    // arrow functions and other things
    {code: 'doThing().then(() => 4)', parserOptions: {ecmaVersion: 6}},
    {code: 'doThing().then(() => { throw 4 })', parserOptions: {ecmaVersion: 6}},
    {code: 'doThing().then(()=>{}, () => 4)', parserOptions: {ecmaVersion: 6}},
    {code: 'doThing().then(()=>{}, () => { throw 4 })', parserOptions: {ecmaVersion: 6}},
    {code: 'doThing().catch(() => 4)', parserOptions: {ecmaVersion: 6}},
    {code: 'doThing().catch(() => { throw 4 })', parserOptions: {ecmaVersion: 6}},

    // random functions and callback methods
    'var x = function() { return Promise.resolve(4) }',
    'function y() { return Promise.resolve(4) }',
    'function then() { return Promise.reject() }',
    'doThing(function(x) { return Promise.reject(x) })'
  ],

  invalid: [

    // wrapped resolve is bad
    {code: 'doThing().then(function() { return Promise.resolve(4) })', errors: [{message: resolveMessage}]},
    {code: 'doThing().then(null, function() { return Promise.resolve(4) })', errors: [{message: resolveMessage}]},
    {code: 'doThing().catch(function() { return Promise.resolve(4) })', errors: [{message: resolveMessage}]},

    // wrapped reject is bad
    {code: 'doThing().then(function() { return Promise.reject(4) })', errors: [{message: rejectMessage}]},
    {code: 'doThing().then(null, function() { return Promise.reject(4) })', errors: [{message: rejectMessage}]},
    {code: 'doThing().catch(function() { return Promise.reject(4) })', errors: [{message: rejectMessage}]},

    // needs to also look at weird paths
    {
      code: 'doThing().then(function(x) { if (x>1) { return Promise.resolve(4) } else { throw "bad" } })',
      errors: [{message: resolveMessage}]
    },
    {code: 'doThing().then(function(x) { if (x>1) { return Promise.reject(4) } })', errors: [{message: rejectMessage}]},
    {
      code: 'doThing().then(null, function() { if (true && false) { return Promise.resolve() } })',
      errors: [{message: resolveMessage}]
    },
    {code: 'doThing().catch(function(x) { return x && Promise.resolve(4) })', errors: [{message: resolveMessage}]},
    {code: 'doThing().catch(function(x) { return x && Promise.reject(4) })', errors: [{message: rejectMessage}]}

  ]
})
```

**Comment by xjamundx** [comment]
Should land on Monday...
